crates/bevy_math/src/smooth.rs

```diff
@@ -0,0 +1,118 @@
+use crate::{Vec2, Vec3};
+
+/// Smooths value to a goal using a damped spring.
+pub trait SmoothDamp {
+    /// Smooths value to a goal using a damped spring.
+    ///
+    /// `smooth_time` is the expected time to reach the target when at maximum velocity.
+    ///
+    /// Returns smoothed value and new velocity.
+    fn smooth_damp(
+        from: Self,
+        to: Self,
+        velocity: Self,
+        smooth_time: f32,
+        delta_time: f32,
+    ) -> (Self, Self)
```

Comment:
IMO returning structs > return tuples.
Having a struct makes it much clearer what each value is and prevent bugs where someone accidentally switches `.0` and `.1`.

So having this return something like
```rust
struct SmoothDampResult<T> {
    smoothed: T,
    velocity: T,
}
```
would be much clearer and prevent a group a user-end bugs.

Reply:
At the moment, rust doesn't support tuple deconstruction over existing variables. But in the future, I expect users to call the function this way: `(obj.pos, obj.vel) = smooth_damp(obj.pos, target, obj.vel, 0.1, dt);` That's the only reason I wrote it this way. 

If I were to use a SmoothDamp structure, I would use it as input too. 

The second option would be to make the velocity input &mut, like C/C++/C# version do, but I fear needless lifetime issues and I like the function being pure...

Reply:
Input parameters via `&mut` are not a good option IMO. 

I would still argue for the `struct` return value since you can do something like this:

```rust
        let SmoothDampResult { smoothed, velocity } = Vec3::smooth_damp(
            player_move.smoothed_input,
            raw_input,
            player_move.smoothed_input_velocity,
            0.15,
            time.delta_seconds(),
        );
        player_move.smoothed_input = smoothed;
        player_move.smoothed_input_velocity = velocity ;
```

But I would leave this decision up to @cart :smile: